### DISCUSSION TITLE
Where are the VS Code extension secrets stored exactly?

### DISCUSSION THREAD

**Discussion by ykdojo** [microsoft/vscode-discussions#748] [Extension Development QnA] [answered]
When using [SecretStorage in VS Code API](https://code.visualstudio.com/api/references/vscode-api#SecretStorage) through an extension, where are the secrets stored, and is there a way for the user to view and edit them?

I saw this from @TylerLeonhardt:

> Secrets are stored in the keyring of the OS you're on.
> 
> We use [http://github.com/atom/node-keytar](https://github.com/atom/node-keytar)
> 
> A native Node module to get, add, replace, and delete passwords in system's keychain. On macOS the passwords are managed by the Keychain, on Linux they are managed by the Secret Service API/libsecret, and on Windows they are managed by Credential Vault.
> 
> Additionally we encrypt the strings before adding them to the keyring to make it harder for other processes and extensions that also have access to the keyring to read the secrets stored with the SecretStorage API.

However, I was not able to identify where it is stored on my system (Mac Keychain).

**Comment by TylerLeonhardt** [comment, marked as answer]
That's an old statement of mine. Now we use https://www.electronjs.org/docs/latest/api/safe-storage for encrypting secrets. This method, on macOS at least, creates an encryption key and stores it in Keychain Access under `Code Safe Storage`, or `Code - Insiders Safe Storage`.

We use this API to encrypt your secrets, and then once they're encrypted we store them in the user data directory which is a SQLite DB on the filesystem where everything else related to VS Code state and whatnot are stored. Hope that helps!

**Reply by ohadschn** [reply]
I wanted to understand where secrets were stored on Windows, so I followed the code -

It looks like electron's `safeStorage` in turn uses Chromium's OSCrypt: https://github.com/electron/electron/blob/3609fc7402881b1d51f6c56249506d5dd3fcbe93/shell/browser/api/electron_api_safe_storage.cc#L69

Where the code for OSCrypt can be found here:
https://chromium.googlesource.com/chromium/src/+/lkgr/components/os_crypt/sync/

Specifically:
 * Windows uses DPAPI: https://chromium.googlesource.com/chromium/src/+/lkgr/components/os_crypt/sync/os_crypt_win.cc
 * Mac uses the Keychain: https://chromium.googlesource.com/chromium/src/+/lkgr/components/os_crypt/sync/os_crypt_mac.mm
 * Linux uses some abstraction called `KeyStorageLinux` which maps to an OS-level library (e.g `libsecret`), which I'm guessing would mean keyring in most cases: https://chromium.googlesource.com/chromium/src/+/lkgr/components/os_crypt/sync/os_crypt_linux.cc
 * On POSIX, apparently they use "peanut-based" security: https://chromium.googlesource.com/chromium/src/+/lkgr/components/os_crypt/sync/os_crypt_posix.cc (seriously, search for `peanuts` in the code...)

**Comment by cgrama** [comment]
@TylerLeonhardt 
Few more secretstorage api issues/questions at https://github.com/microsoft/vscode-discussions/discussions/1188
